packages/lit-html/src/lib/lit-html.ts

```diff
@@ -12,9 +12,548 @@
+    // we also dynamically create a regex to find the matching end tags for raw
+    // text elements.
+
+    const textRegex = /<((?:!--)|(?:\w*))/g;
+    const commentRegex = /-->/g;
+    const tagRegex = /(>)|(?:(?:[ \x09\x0a\x0c\x0d])([^\0-\x1F\x7F-\x9F "'>=/]+)([ \x09\x0a\x0c\x0d]*=[ \x09\x0a\x0c\x0d]*([^ \x09\x0a\x0c\x0d"'`<>=]|("|')|$)))/g;
+    const singleQuoteAttr = /'/g;
+    const doubleQuoteAttr = /"/g;
+
+    let html = type === 2 ? '<svg>' : '';
+    let quote: string | undefined = undefined;
+    let node: Node | null;
+    let nodeIndex = 0;
+    let bindingIndex = 0;
+
+    // The current parsing state, represented as a reference to one of the
+    // regexes
+    let regex = textRegex;
+
+    for (let i = 0; i < strings.length - 1; i++) {
+      const s = strings[i];
+      let attrNameEnd = -1;
+      let lastIndex = 0;
+
+      // The conditions in this loop handle the current parse state, and the
+      // assignments to the `regex` variable are the state transitions.
+      while (lastIndex < s.length) {
+        // Make sure we start searching from where we previously left off
+        regex.lastIndex = lastIndex;
+        const match = regex.exec(s);
+        if (match === null) {
+          if (quote === undefined) {
+            attrNameEnd = -1;
+          }
+          break;
+        }
+        lastIndex = regex.lastIndex;
+        if (regex === textRegex) {
+          // TODO (justinfagnani): there are more ways to open a comment in
+          // HTML, like `<?` or `</{`. Figure out whether / how to handle that.
+          if (match[1] === '!--') {
+            regex = commentRegex;
+          } else {
+            if (rawTextElement.test(match[1])) {
+              regex = new RegExp(`<\/${match[1]}`);
+            } else {
+              regex = tagRegex;
+            }
+          }
+        } else if (regex === commentRegex) {
+          regex = textRegex;
+        } else if (
+          regex === tagRegex ||
+          regex === doubleQuoteAttr ||
+          regex === singleQuoteAttr
+        ) {
+          if (match[1] === '>') {
+            regex = textRegex;
+          } else if (quote && (match[0] === '"' || match[0] === "'")) {
+            // End quoted attribute
+            attrNameEnd = -1;
+            regex = tagRegex;
+          } else {
+            attrNameEnd = regex.lastIndex - match[3].length;
+            quote = match[5];
+            regex =
+              quote === undefined
+                ? tagRegex
+                : quote === '"'
+                ? doubleQuoteAttr
+                : singleQuoteAttr;
+          }
+        } else {
+          // Not one of the five state regexes, so we're at the close of a raw a
+          // text element.
+          regex = tagRegex;
+        }
+      }
+
+      html +=
+        regex === textRegex
+          ? s + nodeMarker
+          : (attrNameEnd === -1
+              ? s
```

Comment:
It does have a marker inserted, this is a nested ternary. A little hard to read. I did that to only reference `marker` once. It's clearer in #1212 where I use an if statement since we needed record the attribute name on the attribute case too.